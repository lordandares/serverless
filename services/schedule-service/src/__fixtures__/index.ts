import { Context } from 'aws-lambda'

export const areaDocument = {
  _id: 'location1',
  address: {
    city: 'Bethesda',
    country: 'US',
    postal: '20889',
    state: 'MD',
    street: '8901 Rockville Pike',
    street2: '',
  },
  altName: 'WRNMMC Facility',
  application: '56789671d4c6488b128c7f9c',
  areaSize: 536559.55514508,
  deleted: false,
  floorsRef: [0],
  groupType: 'location',
  keywords: ['hospital'],
  name: 'WRNMMC Facility',
  serviceHours: {
    hours: [
      {
        description: 'MON 00:00 - TUE 00:00',
        duration: 86400000,
        end: 86400000,
        start: 0,
        type: 'DEFAULT',
      },
    ],
    timezone: 'America/New_York',
  },
  tags: [],
  timezone: 'America/New_York',
}

export const occurrenceDocument = {
  applicationId: 'application1',
  createdAt: '2019-01-01T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    occurrenceInterval: [1, 2],
    scheduleName: 'Schedule 1',
    serviceInterval: [1, 2],
    timezone: 'Australia/Melbourne',
    type: 'visit',
  },
  endAt: '2019-02-01T23:59:59.999Z',
  groupType: 'occurrence',
  locationId: '5d163137b8b3b7000127edd1',
  location_endAt_occurrenceId: 'location1-2019-02-02T00:00:00.000Z-occurrence1',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-01-01T00:00-occurrence1',
  startAt: '2019-02-01T00:00:00.000Z',
  status: 'pending',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}

export const occurrenceTimerDocument = {
  pk: `timer-2019-01-01T00:00`,
  sk: `timer#application1-occurrence#2019-01-01T00:00-occurrence`,
  endAt: '2019-01-01T23:59:59.999Z',
  groupType: `occurrence-timer`,
  scheduleId: 'schedule1',
  targetArn: 'someArn',
}

export const scheduleDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    areas: ['area1'],
    enabled: true,
    included: {},
    locations: ['location1'],
    name: 'Schedule One',
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
    strategy: {
      options: {
        duration: {
          unit: 'minute',
          value: 15,
        },
        frequency: {
          unit: 'minute',
          value: 15,
        },
      },
      type: 'stopwatch',
    },
    type: 'visit',
  },
  endAt: '2019-10-1500:00:00.000Z',
  groupType: 'schedule',
  pk: 'application1-schedule',
  scheduleId: 'schedule1',
  sk: '2019-10-13T00:00:00.000Z-schedule1',
  startAt: '2019-10-1400:00:00.000Z',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}

export const scheduleLocationDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: '123456789',
    label: 'label',
    type: 'system',
  },
  data: {
    name: 'Schedule One',
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
  },
  groupType: 'location',
  pk: 'application1-location',
  schedules: ['schedule1'],
  sk: 'location1',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: '123456789',
    label: 'label',
    type: 'system',
  },
}

export const schedulePayload = {
  areas: ['area1'],
  enabled: true,
  endAt: '2019-10-1500:00:00.000Z',
  locations: ['location1'],
  name: 'Schedule One',
  serviceHours: {
    hours: [
      {
        description: 'MON 11:00 - TUE 00:00',
        end: 86400000,
        start: 39600000,
        type: 'DEFAULT',
      },
    ],
    timezone: 'Australia/Melbourne',
  },
  startAt: '2019-10-1400:00:00.000Z',
  strategy: {
    options: {
      duration: {
        unit: 'minute',
        value: 15,
      },
      frequency: {
        unit: 'minute',
        value: 15,
      },
    },
    type: 'stopwatch',
  },
  type: 'visit',
}

export const rulePayload = {
  applicationId: 'application-id-1111',
  data: {
    occurrenceInterval: [1580761454998, 1580762054997],
  },
  hook: {
    endpoint: 'sns:endpoint:uri',
    type: 'sns',
  },
  locationId: 'location-id-11111',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-06-02T00:00:00.000Z-occurrence1',
  startAt: '2019-02-01T00:00:00.000Z',
  type: 'visit',
}

export const rulePatternDocument = {
  createdAt: '2019-06-02T00:00:00.000Z',
  createdBy: {
    id: 'created-by-id-11111',
    label: 'created-by-label-1111',
    type: 'created-by-type-11111',
  },
  groupType: 'rule-pattern',
  matches: {
    occurrence1: {
      pk: 'application1-occurrence',
      sk: '2019-06-01T00:00:00.000Z-occurrence1',
    },
  },
  pk: 'rule-pattern-application-id-11111-location-id-11111',
  sk: 'visit',
  updatedAt: '2019-06-02T00:00:00.000Z',
}

export const context: Context = {
  awsRequestId: 'aws-request-id',
  callbackWaitsForEmptyEventLoop: false,
  clientContext: undefined,
  done: jest.fn(),
  fail: jest.fn(),
  functionName: 'function-name',
  functionVersion: 'function-version',
  getRemainingTimeInMillis: jest.fn(),
  identity: undefined,
  invokedFunctionArn: 'invoked-function-arn',
  logGroupName: 'log-group-name',
  logStreamName: 'log-stream-name',
  memoryLimitInMB: 0,
  succeed: jest.fn(),
}
