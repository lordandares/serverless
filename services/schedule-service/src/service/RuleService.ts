import { errors, schemas } from '@lighthouse/serverless-common'
import { DynamoRepository } from '../repository/dynamo'
import { payloadToRulePatternDocument } from './lib/ruleTransform'

declare var process: {
  env: {
    TABLE_SCHEDULES: string
  }
}

const removeRule = async (id: string): Promise<void> => {
  if (!id) {
    throw new errors.ApplicationError({
      message: `The rule \`id\` of the resource to delete is missing`,
    })
  }

  console.debug('RemoveRule: id', { id })

  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)
  // await dbObject.remove(id)
}

const RuleService = {
  removeRule,
}

export { RuleService }
