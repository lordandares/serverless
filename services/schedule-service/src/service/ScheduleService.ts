import { scheduling } from '@lighthouse/common'
import { errors, mongo, schemas } from '@lighthouse/serverless-common'
import { SNS, StepFunctions } from 'aws-sdk'
import { DocumentClient as DC } from 'aws-sdk/clients/dynamodb'
import * as BPromise from 'bluebird'
import {
  filter,
  first,
  get,
  isEmpty,
  map,
  toString,
  uniq,
  without,
} from 'lodash'
import * as moment from 'moment'
import { v4 as uuid } from 'uuid'

import { OCCURRENCE_STATUS_PENDING } from '../constants'
import { DynamoRepository } from '../repository/dynamo'
import {
  documentToSchedulePayload,
  payloadToScheduleDocument,
} from '../service/lib/transform'

declare var process: {
  env: {
    CREATE_SCHEDULE_STEP_FUNCTION: string
    OCCURRENCE_EXPIRED_ARN: string
    TABLE_SCHEDULES: string
    UPDATE_SCHEDULE_STEP_FUNCTION: string
  }
}

interface Document {
  pk: string
  sk: string
}

interface CreateSchedule {
  applicationId: string
  body: schemas.SchedulePayloadSchema
  userId: string
}

const createSchedule = async ({
  applicationId,
  body,
  userId,
}: CreateSchedule): Promise<schemas.SchedulePayloadSchema> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)
  const sfs = new StepFunctions()

  if (!applicationId) {
    throw new errors.ApplicationError({
      message: `The \`applicationId\` is missing from headers`,
    })
  }

  if (!userId) {
    throw new errors.ApplicationError({
      message: `The \`userId\` is missing from headers`,
    })
  }

  if (isEmpty(body)) {
    throw new errors.ValidationError({
      message: `A schedule body is required`,
    })
  }

  console.debug('CreateSchedule: body', { body })

  await schemas.validate({
    data: body,
    schema: schemas.schedulePayloadSchema,
  })

  const { locations } = body

  const includedLocations = await BPromise.reduce(
    locations,
    async (accum, id) => {
      const location = await getArea({ applicationId, locationId: id })

      if (!location) {
        return accum
      }

      const { name } = location

      accum[id] = { name }
      return accum
    },
    {},
  )

  console.debug('CreateSchedule: included locations', { includedLocations })

  const included = { locations: includedLocations }

  const additional = { applicationId, included, userId }

  const document: schemas.ScheduleDocumentSchema = payloadToScheduleDocument({
    payload: body,
    additional,
  })

  try {
    await schemas.validate({
      data: document,
      options: { context: { isNew: true } },
      schema: schemas.scheduleDocumentSchema,
    })
  } catch (err) {
    throw new errors.ApplicationError({ message: err.message })
  }

  const { scheduleId } = document

  console.debug('CreateSchedule: document to be created', { document })

  const scheduleDocument: schemas.ScheduleDocumentSchema = await dbObject.put(
    document,
  )

  console.debug('CreateSchedule: invoking step function', {
    applicationId,
    locations,
    scheduleId,
    stateMachineArn: process.env.CREATE_SCHEDULE_STEP_FUNCTION,
  })

  await sfs
    .startExecution({
      input: JSON.stringify({
        applicationId,
        locations,
        scheduleId,
      }),
      name: `${applicationId}-${scheduleId}-${Date.now()}`,
      stateMachineArn: process.env.CREATE_SCHEDULE_STEP_FUNCTION,
    })
    .promise()

  const responsePayload: schemas.SchedulePayloadSchema = documentToSchedulePayload(
    {
      document: scheduleDocument,
    },
  )

  console.debug('CreateSchedule: response payload', { responsePayload })

  return responsePayload
}

const deleteOccurrences = async (
  scheduleId: string,
): Promise<DC.TransactWriteItemsOutput | []> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)

  if (!scheduleId) {
    throw new errors.ApplicationError({
      message: 'The schedule `id` of the resource to delete is missing',
    })
  }

  const occurrences = (await dbObject.getBySchedule({
    groupType: 'occurrence',
    scheduleId,
    status: 'pending',
  })) as schemas.ScheduleOccurrenceDocumentSchema[]

  const items = map(occurrences, occurrence => ({
    pk: occurrence.pk,
    sk: occurrence.sk,
  }))

  console.debug('DeleteOccurrences: items to be deleted', { items })

  const results = await dbObject.remove(items)

  return results
}

interface GetArea {
  applicationId: string
  locationId: string
}

async function getArea({ applicationId, locationId }: GetArea) {
  const areasCollection = await mongo.getCollection('areas')

  const area = await areasCollection.findOne({
    _id: new mongo.ObjectId(toString(locationId)),
    application: new mongo.ObjectId(toString(applicationId)),
  })

  console.debug('GetArea: returned area', { applicationId, area, locationId })

  return area
}

interface GetSchedule {
  applicationId: string
  scheduleId?: string
}

const getSchedule = async ({
  applicationId,
  scheduleId,
}: GetSchedule): Promise<schemas.SchedulePayloadSchema> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)

  if (!applicationId) {
    throw new errors.ApplicationError({
      message: `The \`applicationId\` is missing from headers`,
    })
  }

  if (!scheduleId) {
    throw new errors.ValidationError({
      message: `The schedule \`id\` is required`,
    })
  }

  const result = (await dbObject.getBySchedule({
    groupType: 'schedule',
    limit: 1,
    scheduleId,
  })) as schemas.ScheduleDocumentSchema[]

  if (isEmpty(result)) {
    throw new errors.ResourceNotFoundError({
      id: scheduleId,
      resource: 'Schedule',
    })
  }

  const document: schemas.ScheduleDocumentSchema = result[0]
  const responsePayload: schemas.SchedulePayloadSchema = documentToSchedulePayload(
    {
      document,
    },
  )

  return responsePayload
}

interface ListSchedules {
  applicationId: string
}

const listSchedules = async ({
  applicationId,
}: ListSchedules): Promise<schemas.SchedulePayloadSchema[]> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)

  if (!applicationId) {
    throw new errors.ApplicationError({
      message: `The \`applicationId\` is missing from headers`,
    })
  }

  const scheduleDocuments = (await dbObject.listSchedules(
    applicationId,
  )) as schemas.ScheduleDocumentSchema[]

  const responsePayload: schemas.SchedulePayloadSchema[] = map(
    scheduleDocuments,
    scheduleDocument =>
      documentToSchedulePayload({
        document: scheduleDocument,
      }),
  )

  return responsePayload
}

interface RemoveSchedule {
  applicationId: string
  scheduleId: string
}

const removeSchedule = async ({
  applicationId,
  scheduleId,
}: RemoveSchedule): Promise<void> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)

  if (!applicationId) {
    throw new errors.ApplicationError({
      message: `The \`applicationId\` is missing from headers`,
    })
  }

  if (!scheduleId) {
    throw new errors.ValidationError({
      message: 'The schedule `id` of the resource to delete is missing',
    })
  }

  const items = (await dbObject.getBySchedule({
    scheduleId,
  })) as Array<
    schemas.ScheduleDocumentSchema | schemas.ScheduleOccurrenceDocumentSchema
  >

  console.debug('RemoveSchedule: database results', { results: items })

  const itemKeys = items.map(({ pk, sk }) => ({ pk, sk }))

  console.debug('RemoveSchedule: items to be removed', { items: itemKeys })

  await dbObject.remove(itemKeys)

  const schedule = first(
    filter(items, { groupType: 'schedule' }),
  ) as schemas.ScheduleDocumentSchema

  if (!schedule) {
    return
  }

  const { locations } = schedule.data

  console.debug('RemoveSchedule: schedule locations', { locations })

  await BPromise.each(locations, async (locationId: string) => {
    const locationDocument = (await dbObject.get({
      pk: `${applicationId}-location`,
      sk: locationId,
    })) as schemas.ScheduleLocationDocumentSchema

    console.debug('RemoveSchedule: location database result', {
      locationDocument,
    })

    if (!locationDocument) {
      console.info('RemoveSchedule: Schedule referenced a missing location', {
        locationId,
        scheduleId,
      })

      return
    }

    const { schedules } = locationDocument

    const updatedSchedules = without(schedules, scheduleId)

    const updatedLocationDocument = {
      ...locationDocument,
      schedules: updatedSchedules,
    }

    console.debug('RemoveSchedule: location document update', {
      updatedLocationDocument,
    })

    await dbObject.put(updatedLocationDocument)
  })
}

interface UpdateSchedule {
  applicationId: string
  body: schemas.SchedulePayloadSchema
  scheduleId: string | null
  userId: string
}

const updateSchedule = async ({
  applicationId,
  body,
  scheduleId,
  userId,
}: UpdateSchedule): Promise<schemas.SchedulePayloadSchema> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)
  const sfs = new StepFunctions()

  if (!applicationId) {
    throw new errors.ApplicationError({
      message: `The \`applicationId\` is missing from headers`,
    })
  }

  if (!userId) {
    throw new errors.ApplicationError({
      message: `The \`userId\` is missing from headers`,
    })
  }

  if (!scheduleId) {
    throw new errors.ValidationError({
      message: `The schedule \`id\` is required`,
    })
  }

  if (isEmpty(body)) {
    throw new errors.ValidationError({
      message: `A schedule body is required`,
    })
  }

  console.debug('UpdateSchedule: body', { body })

  await schemas.validate({
    data: body,
    schema: schemas.schedulePayloadSchema,
  })

  const result = (await dbObject.getBySchedule({
    groupType: 'schedule',
    limit: 1,
    scheduleId,
  })) as schemas.ScheduleDocumentSchema[]

  console.debug('UpdateSchedule: database result', { result })

  if (isEmpty(result)) {
    throw new errors.ResourceNotFoundError({
      id: scheduleId,
      resource: 'Schedule',
    })
  }

  const existingDocument: schemas.ScheduleDocumentSchema = result[0]

  console.debug('UpdateSchedule: existing document', { existingDocument })

  const existingDocumentPayload: schemas.ScheduleResponseSchema = documentToSchedulePayload(
    {
      document: existingDocument,
    },
  )

  console.debug('UpdateSchedule: existing document payload', {
    existingDocumentPayload,
  })

  const { locations } = body

  const includedLocations = await BPromise.reduce(
    locations,
    async (accum, id) => {
      const location = await getArea({ applicationId, locationId: id })

      if (!location) {
        return accum
      }

      const { name } = location

      accum[id] = { name }
      return accum
    },
    {},
  )

  console.debug('UpdateSchedule: included locations', { includedLocations })

  const included = { locations: includedLocations }

  const payload = {
    ...existingDocumentPayload,
    ...body,
  }

  const additional = {
    applicationId,
    createdAt: existingDocument.createdAt,
    createdBy: existingDocument.createdBy,
    id: scheduleId,
    included,
    userId,
  }

  const document: schemas.ScheduleDocumentSchema = payloadToScheduleDocument({
    payload,
    additional,
  })

  console.debug('UpdateSchedule: payload to document', { document })

  try {
    await schemas.validate({
      data: document,
      options: { context: { isNew: false } },
      schema: schemas.scheduleDocumentSchema,
    })
  } catch (err) {
    throw new errors.ApplicationError({ message: err.message })
  }

  console.debug('UpdateSchedule: document to be updated', { document })

  const updatedDocument: schemas.ScheduleDocumentSchema = await dbObject.put(
    document,
  )

  console.debug('UpdateSchedule: invoking step function', {
    applicationId,
    locations,
    scheduleId,
    stateMachineArn: process.env.UPDATE_SCHEDULE_STEP_FUNCTION,
  })

  await sfs
    .startExecution({
      input: JSON.stringify({
        applicationId,
        locations,
        scheduleId,
      }),
      name: `${applicationId}-${scheduleId}-${Date.now()}`,
      stateMachineArn: process.env.UPDATE_SCHEDULE_STEP_FUNCTION,
    })
    .promise()

  const responsePayload: schemas.ScheduleResponseSchema = documentToSchedulePayload(
    {
      document: updatedDocument,
    },
  )

  console.debug('UpdateSchedule: response payload', { responsePayload })

  return responsePayload
}

interface UpsertLocation {
  applicationId: string
  arn: string
  arnId: string
  locationId: string
  scheduleId?: string
}

const upsertLocation = async ({
  applicationId,
  arn,
  arnId,
  locationId,
  scheduleId,
}: UpsertLocation): Promise<schemas.ScheduleLocationDocumentSchema> => {
  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)

  if (!applicationId) {
    throw new errors.ApplicationError({
      message: 'The `applicationId` is missing',
    })
  }

  if (!arn) {
    throw new errors.ApplicationError({
      message: 'The `arn` is missing',
    })
  }

  if (!arnId) {
    throw new errors.ApplicationError({
      message: 'The `arnId` is missing',
    })
  }

  if (!locationId) {
    throw new errors.ApplicationError({
      message: 'The `locationId` is missing',
    })
  }

  const area = await getArea({ applicationId, locationId })

  console.debug('UpsertLocation: area', { area })

  const { name, serviceHours } = area

  const existingLocation = await dbObject.get({
    pk: `${applicationId}-location`,
    sk: locationId,
  })

  console.debug('UpsertLocation: existing location', { existingLocation })

  const actor: schemas.ActorSchema = { id: arnId, label: arn, type: 'system' }

  const now: string = moment
    .utc()
    .toDate()
    .toISOString()

  const createdAt = existingLocation ? existingLocation.createdAt : now
  const createdBy = existingLocation ? existingLocation.createdBy : actor
  const schedules =
    existingLocation && scheduleId
      ? uniq([...existingLocation.schedules, scheduleId])
      : existingLocation && !scheduleId
        ? existingLocation.schedules
        : !existingLocation && scheduleId
          ? [scheduleId]
          : []

  const document: schemas.ScheduleLocationDocumentSchema = {
    applicationId,
    createdAt,
    createdBy,
    data: { name, serviceHours },
    groupType: 'location',
    pk: `${applicationId}-location`,
    schedules,
    sk: locationId,
    updatedAt: now,
    updatedBy: actor,
  }

  try {
    await schemas.validate({
      data: document,
      schema: schemas.scheduleLocationDocumentSchema,
    })
  } catch (err) {
    throw new errors.ApplicationError({ message: err.message })
  }

  console.debug('UpsertLocation: document to be created', { document })

  const result = (await dbObject.put(
    document,
  )) as schemas.ScheduleLocationDocumentSchema

  if (existingLocation && !isEmpty(existingLocation.schedules)) {
    await BPromise.each(existingLocation.schedules, async (id: string) => {
      const scheduleResult = (await dbObject.getBySchedule({
        groupType: 'schedule',
        limit: 1,
        scheduleId: id,
      })) as schemas.ScheduleDocumentSchema[]

      console.debug('UpsertLocation: schedule database result', {
        scheduleResult,
      })

      if (isEmpty(scheduleResult)) {
        console.debug(
          'UpsertLocation: Location references a missing schedule',
          {
            locationId,
            scheduleId: id,
          },
        )

        return
      }

      const existingDocument = scheduleResult[0]

      console.debug('UpsertLocation: schedule', { schedule: existingDocument })

      const nextSchedule: schemas.ScheduleDocumentSchema = {
        ...existingDocument,
        data: {
          ...existingDocument.data,
          included: {
            ...existingDocument.data.included,
            locations: {
              ...existingDocument.data.included.locations,
              [locationId]: { name },
            },
          },
        },
      }

      console.debug('UpsertLocation: document to be updated', {
        document: nextSchedule,
      })

      await dbObject.put(nextSchedule)
    })
  }

  return result
}

const ScheduleService = {
  createSchedule,
  deleteOccurrences,
  getSchedule,
  listSchedules,
  removeSchedule,
  updateSchedule,
  upsertLocation,
}

export { ScheduleService }
