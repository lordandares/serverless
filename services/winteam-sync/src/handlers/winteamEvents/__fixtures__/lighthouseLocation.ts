const lighthouseLocation = {
  _id: 'objectId',
  name: 'Powell Motors',
  type: 'location',
  address: {
    street: 'Center',
    street2: '7855 So. River Pkwy',
    city: 'San Diego',
    state: 'CA',
    postalCode: '852840000',
    country: 'US',
  },
  geometry: {
    type: 'Polygon',
    coordinates: [
      [
        [39.984, -75.34310067963625],
        [39.98365162727836, -75.34310500985033],
        [39.9833066089773, -75.34311795879778],
        [39.98296826723548, -75.3431394017952],
        [39.982639859937834, -75.34316913237075],
        [39.98232454936074, -75.34320686425103],
        [39.98202537173541, -75.3432522341163],
        [39.98174520802207, -75.34330480509715],
        [39.98148675617558, -75.34336407097913],
        [39.98125250516942, -75.34342946107526],
        [39.9810447110278, -75.34350034571895],
        [39.98086537509659, -75.34357604232524],
        [39.980716224762624, -75.34365582196158],
        [39.980598696807064, -75.34373891636542],
        [39.98051392355378, -75.34382452534051],
        [39.98046272194656, -75.34391182446154],
        [39.98044558566106, -75.3439999730124],
        [39.98046268032836, -75.34408812208183],
        [39.98051384191675, -75.3441754227386],
        [39.980598578288465, -75.34426103420762],
        [39.98071607391706, -75.3443441319677],
        [39.98086519772097, -75.34442391569367],
        [39.98104451393857, -75.34449961696573],
        [39.981252295940614, -75.34457050667213],
        [39.98148654284773, -75.34463590203326],
        [39.98174499879326, -75.34469517318026],
        [39.98202517464618, -75.34474774922379],
        [39.98232437198512, -75.34479312375487],
        [39.982639709092275, -75.34483085972474],
        [39.98296814871688, -75.34486059365653],
        [39.98330652734027, -75.34488203914788],
        [39.98365158566017, -75.34489498963106],
        [39.984, -75.34489932036371],
        [39.98434841433984, -75.34489498963106],
        [39.984693472659735, -75.34488203914788],
        [39.985031851283125, -75.34486059365653],
        [39.98536029090773, -75.34483085972474],
        [39.98567562801488, -75.34479312375487],
        [39.98597482535382, -75.34474774922379],
        [39.98625500120675, -75.34469517318026],
        [39.98651345715227, -75.34463590203326],
        [39.98674770405939, -75.34457050667213],
        [39.98695548606143, -75.34449961696573],
        [39.98713480227903, -75.34442391569367],
        [39.987283926082945, -75.3443441319677],
        [39.987401421711546, -75.34426103420762],
        [39.98748615808325, -75.3441754227386],
        [39.987537319671645, -75.34408812208183],
        [39.987554414338945, -75.3439999730124],
        [39.98753727805345, -75.34391182446154],
        [39.987486076446224, -75.34382452534051],
        [39.98740130319294, -75.34373891636542],
        [39.98728377523738, -75.34365582196158],
        [39.98713462490341, -75.34357604232524],
        [39.98695528897221, -75.34350034571895],
        [39.98674749483058, -75.34342946107526],
        [39.986513243824426, -75.34336407097913],
        [39.98625479197794, -75.34330480509715],
        [39.98597462826459, -75.3432522341163],
        [39.98567545063927, -75.34320686425103],
        [39.98536014006217, -75.34316913237075],
        [39.98503173276453, -75.3431394017952],
        [39.9846933910227, -75.34311795879778],
        [39.984348372721634, -75.34310500985033],
        [39.984, -75.34310067963625],
      ],
    ],
  },
  plugins: {
    winteam: {
      options: {
        jobId: '947049c7-e118-4ee9-8e2e-c956285917f1',
        jobNumber: '50012',
      },
    },
  },
  timezone: 'America/Los_Angeles',
}

export default lighthouseLocation
