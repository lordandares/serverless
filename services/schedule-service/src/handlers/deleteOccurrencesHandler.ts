import { errors } from '@lighthouse/serverless-common'

import { ScheduleService } from '../service/ScheduleService'

interface DeleteOccurrencesHandler {
  scheduleId: string
}

export async function deleteOccurrencesHandler(
  event: DeleteOccurrencesHandler,
) {
  try {
    if (!event) {
      throw new errors.ApplicationError({
        message: 'DeleteOccurrencesHandler: missing event',
      })
    }

    const { scheduleId } = event
    const results = await ScheduleService.deleteOccurrences(scheduleId)

    return results
  } catch (err) {
    console.error('DeleteOccurrencesError', {
      err,
      event,
    })

    if (errors.isKnownError(err)) {
      throw err
    }

    throw new errors.UnknownError()
  }
}
