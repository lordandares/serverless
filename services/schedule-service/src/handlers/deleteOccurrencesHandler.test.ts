import { mongo, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk-mock'

import { occurrenceDocument } from '../__fixtures__'
import { ScheduleService } from '../service/ScheduleService'
import { deleteOccurrencesHandler } from './deleteOccurrencesHandler'

beforeEach(() => {
  process.env.TABLE_SCHEDULES = 'table-schedules'
})

test('application error returned when event', async () => {
  expect.assertions(1)

  const event = null

  try {
    await deleteOccurrencesHandler(event)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: DeleteOccurrencesHandler: missing event]`,
    )
  }
})

test('application error returned when missing scheduleId', async () => {
  expect.assertions(1)

  const event = {
    scheduleId: null as any,
  }

  try {
    await deleteOccurrencesHandler(event)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: The schedule \`id\` of the resource to delete is missing]`,
    )
  }
})

test('application error returned when querying for occurrences errors', async () => {
  expect.assertions(1)

  const occurrences = [{ ...occurrenceDocument }]

  const querySpy = jest.fn().mockRejectedValue(new Error('Query Error'))

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    scheduleId: 'schedule1',
  }

  try {
    await deleteOccurrencesHandler(event)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(`[ApplicationError: Query Error]`)
  }

  AWS.restore()
})

test('application error returned when removing occurrences errors', async () => {
  expect.assertions(1)

  const occurrences = [{ ...occurrenceDocument }]

  const querySpy = jest.fn().mockResolvedValue({ Items: occurrences })
  const transactWriteSpy = jest
    .fn()
    .mockRejectedValue(new Error('Remove Error'))

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactWriteSpy)

  const event = {
    scheduleId: 'schedule1',
  }

  try {
    await deleteOccurrencesHandler(event)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(`[ApplicationError: Remove Error]`)
  }

  AWS.restore()
})

test('generic error', async () => {
  expect.assertions(1)

  const removeOccurrencesSpy = jest.spyOn(ScheduleService, 'deleteOccurrences')

  removeOccurrencesSpy.mockImplementation(
    jest.fn().mockRejectedValueOnce(new Error('Unknown Error')),
  )

  const event = {
    scheduleId: 'schedule1',
  }

  try {
    await deleteOccurrencesHandler(event)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[UnknownError: Something went wrong! Try again or contact support if the problem persists.]`,
    )
  }

  removeOccurrencesSpy.mockRestore()
})

test('returns no results when schedule has no occurrences', async () => {
  expect.assertions(2)

  const querySpy = jest.fn().mockResolvedValue({ Items: [] })

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    scheduleId: 'schedule1',
  }

  const result = await deleteOccurrencesHandler(event)

  expect(querySpy.mock.calls).toMatchSnapshot()
  expect(result).toEqual([])

  AWS.restore()
})

test('returns removed results when schedule has occurrences', async () => {
  expect.assertions(3)

  const occurrences = [{ ...occurrenceDocument }]

  const querySpy = jest.fn().mockResolvedValue({ Items: occurrences })
  const transactWriteSpy = jest.fn().mockResolvedValue([])

  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)
  AWS.mock('DynamoDB.DocumentClient', 'transactWrite', transactWriteSpy)

  const event = {
    scheduleId: 'schedule1',
  }

  const result = await deleteOccurrencesHandler(event)

  expect(querySpy.mock.calls).toMatchSnapshot()
  expect(transactWriteSpy.mock.calls).toMatchSnapshot()
  expect(result).toEqual([])

  AWS.restore()
})
