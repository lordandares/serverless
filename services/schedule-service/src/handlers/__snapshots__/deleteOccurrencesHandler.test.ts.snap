// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns no results when schedule has no occurrences 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#groupType": "groupType",
        "#scheduleId": "scheduleId",
        "#status": "status",
      },
      "ExpressionAttributeValues": Object {
        ":groupType": "occurrence",
        ":scheduleId": "schedule1",
        ":status": "pending",
      },
      "FilterExpression": "#groupType = :groupType AND #status = :status",
      "IndexName": "ScheduleIdIndex",
      "KeyConditionExpression": "#scheduleId = :scheduleId",
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns removed results when schedule has occurrences 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#groupType": "groupType",
        "#scheduleId": "scheduleId",
        "#status": "status",
      },
      "ExpressionAttributeValues": Object {
        ":groupType": "occurrence",
        ":scheduleId": "schedule1",
        ":status": "pending",
      },
      "FilterExpression": "#groupType = :groupType AND #status = :status",
      "IndexName": "ScheduleIdIndex",
      "KeyConditionExpression": "#scheduleId = :scheduleId",
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns removed results when schedule has occurrences 2`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Delete": Object {
            "Key": Object {
              "pk": "application1-occurrence",
              "sk": "2019-01-01T00:00-occurrence1",
            },
            "TableName": "table-schedules",
          },
        },
      ],
    },
    [Function],
  ],
]
`;
